components/script/serviceworkerjob.rs

```diff
@@ -176,105 +203,158 @@ impl JobQueue {
+                                                         InvokeType::Settle(settle_type),
                                                          &*global);
             script_thread.queue_async_job(box async_job_handler);
         }
-        // TODO Step 7 Update Algorithm
         script_thread.invoke_job_update(job, &*global);
     }
 
     #[allow(unrooted_must_root)]
-    pub fn finish_job(&mut self, scope_url: &Url, global: &GlobalScope, script_thread: &ScriptThread) {
-        if let Some(job_vec) = self.0.get_mut(&scope_url) {
-            let _ = job_vec.remove(0);
+    // https://w3c.github.io/ServiceWorker/#finish-job-algorithm
+    pub fn finish_job(&self, scope_url: Url, global: &GlobalScope, script_thread: &ScriptThread) {
+        if let Some(job_vec) = (*self.0.borrow_mut()).get_mut(&scope_url) {
+            // first conditional causes an early return avoiding the unwrap panic in second conditional
+            if job_vec.first().is_some() && job_vec.first().unwrap().scope_url == scope_url {
+                let _  = job_vec.remove(0);
+            }
             if !job_vec.is_empty() {
                 let promise = Promise::new(global);
-                let run_job_handler = RunJobHandler {
+                let run_job_handler = AsyncJobHandler {
                 global: Trusted::new(global),
                 promise: TrustedPromise::new(promise),
-                scope_url: scope_url.clone()
+                scope_url: scope_url.clone(),
+                invoke_type: InvokeType::Run,
                 };
                 script_thread.queue_run_job(box run_job_handler, global);
             }
         } else {
             warn!("non-existent job vector for url: {:?}", scope_url);
         }
     }
+    // https://w3c.github.io/ServiceWorker/#update-algorithm
+    pub fn update(&self, job: &Job, global: &GlobalScope, script_thread: &ScriptThread) {
+        if let Some(reg) = script_thread.handle_get_registration(&job.scope_url) {
+            // Step 1
+            if reg.get_uninstalling() {
+                let err_type = Error::Type("Update called on an uninstalling registration".to_owned());
+                let async_job_handler = AsyncJobHandler::new(TrustedPromise::new(job.promise.clone()),
+                                                         job.scope_url.clone(),
+                                                         InvokeType::Settle(SettleType::Reject(err_type)),
+                                                         global);
+                return script_thread.queue_async_job(box async_job_handler);
+            }
 
-    pub fn update(&self, global:&GlobalScope, script_thread: &ScriptThread) {}
+            // Step 2
+            if let Some(ref newest_worker) = reg.get_newest_worker() {
+                if (&*newest_worker).get_script_url() == job.script_url /* && job.job_type == JobType::Update*/{
```

Comment:
Ah that, was commented for debugging. That is part of the conditional. updated it.